src/backend/distributed/utils/attribute.c

```diff
@@ -200,59 +207,32 @@ AttributeQueryIfAnnotated(const char *query_string, CmdType commandType)
 
 	if (strncmp(ATTRIBUTE_PREFIX, query_string, strlen(ATTRIBUTE_PREFIX)) == 0)
 	{
-		/* TODO create a function to safely parse the tenant identifier from the query comment */
-		/* query is attributed to a tenant */
-		char *tenantId = (char *) query_string + strlen(ATTRIBUTE_PREFIX);
-		char *tenantEnd = tenantId;
-		while (true && tenantEnd[0] != '\0')
+		char *annotation = ExtractTopComment(query_string);
+		if (annotation != NULL)
 		{
-			if (tenantEnd[0] == ' ' && tenantEnd[1] == '*' && tenantEnd[2] == '/')
+			Datum jsonbDatum = DirectFunctionCall1(jsonb_in, PointerGetDatum(annotation));
+
+			text *tenantIdTextP = ExtractFieldTextP(jsonbDatum, "tId");
+			if (tenantIdTextP != NULL)
 			{
-				break;
+				char *tenantId = UnescapeCommentChars(text_to_cstring(tenantIdTextP));
+				strcpy_s(attributeToTenant, sizeof(attributeToTenant), tenantId);
```

Comment:
We shouldn't error here if the tenant is too long. We should truncate it instead. We should have a test for this too. I tried locally with:
```
psql -p 9700 -c '/*{"tId":"thisisaveryloooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooooongname","cId":1}*/SELECT 1;'
ERROR:  XX000: Memory constraint error: strcpy_s: not enough space for src (errno 406)
LOCATION:  ereport_constraint_handler, citus_safe_lib.c:44
Time: 0.401 ms
```